### PR TITLE
Implement image generators

### PR DESCRIPTION
**Pull request by freekmurze**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -25,19 +25,19 @@
...
     "scrutinizer/ocular": "^1.1",
-    "orchestra/testbench": "3.3.x-dev",
-    "orchestra/database": "3.3.x-dev",
+    "orchestra/testbench": "^3.1",
     "doctrine/dbal": "^2.5.2"
   },
   "conflict": {
-    "php-ffmpeg/php-ffmpeg": "<0.6.1"
+    "php-ffmpeg/php-ffmpeg": "<0.6.1",
+    "spatie/pdf-to-image": "<1.0.0"
```
Comment (sebastiandedeyne) [inline composer.json]: This one isn't necessary

File: src/FileManipulator.php [hunk 2]
```diff
@@ -21,21 +20,25 @@ class FileManipulator
...
-        if ($media->type === Media::TYPE_VIDEO && ! class_exists('\\FFMpeg\\FFMpeg')) {
-            return;
-        }
-
-        if (in_array($media->type, [Media::TYPE_PDF, Media::TYPE_SVG]) && ! class_exists('Imagick')) {
+        $imageGenerator = $media->getImageGenerators()
+            ->map(function (string $imageGeneratorClassName) {
+                return app($imageGeneratorClassName);
+            })
+            ->first(function (ImageGenerator $imageGenerator) use ($media) {
```
Comment (sebastiandedeyne) [inline src/FileManipulator.php]: Will this clash with the 5.3 changes?

File: src/ImageGenerators/BaseGenerator.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+abstract class BaseGenerator implements ImageGenerator
+{
+    public function canConvert(string $path): bool
+    {
+        if (! file_exists($path)) {
+            return false;
+        }
+
+        if (! $this->areRequirementsInstalled()) {
```
Comment (sebastiandedeyne) [inline src/ImageGenerators/BaseGenerator.php]: Rename to `requirementsAreInstalled`?

File: src/ImageGenerators/BaseGenerator.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    public function canHandleExtension(string $extension = ''): bool
+    {
+        return $this->supportedExtensions()->contains($extension);
+    }
+
+    public function getType(): string
+    {
+        return strtolower(class_basename(static::class));
+    }
+}
```
Comment (sebastiandedeyne) [inline src/ImageGenerators/BaseGenerator.php]: Add abstract methods for whatever gets used internally? E.g. `supportedExtensions`, `supportedExtensions`,...

File: src/ImageGenerators/FileTypes/Image.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+use Spatie\MediaLibrary\ImageGenerators\BaseGenerator;
+
+class Image extends BaseGenerator
+{
+    public function convert(string $path, Conversion $conversion = null) : string
+    {
+        return $path;
+    }
+
+    public function areRequirementsInstalled() : bool
```
Comment (sebastiandedeyne) [inline src/ImageGenerators/FileTypes/Image.php]: CS: no space before `:` (StyleCI might/should handle this though)

File: src/ImageGenerators/FileTypes/Image.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        return true;
+    }
+
+    public function supportedExtensions() : Collection
+    {
+        return collect(['png', 'jpg', 'jpeg', 'gif']);
```
Comment (sebastiandedeyne) [inline src/ImageGenerators/FileTypes/Image.php]: I'd opt for plain arrays here since they're less noisy

File: src/Media.php [hunk 7]
```diff
@@ -132,25 +141,15 @@ public function getTypeFromMimeAttribute() : string
...
-        }
-
-        if ($mime === 'image/svg+xml') {
-            return static::TYPE_SVG;
-        }
+        $imageGenerator = $this->getImageGenerators()
+            ->map(function (string $className) {
+                return app($className);
+            })
+            ->first(function (ImageGenerator $imageGenerator) {
```
Comment (sebastiandedeyne) [inline src/Media.php]: L5.3 blues

File: tests/ImageGenerators/VideoTest.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+        $media = $this->testModelWithoutMediaConversions->addMedia($this->getTestWebm())->toMediaLibrary();
+
+        $this->assertTrue($imageGenerator->canConvert($media->getPath()));
+
+        $imageFile = $imageGenerator->convert($media->getPath());
+
+        $this->assertEquals('image/jpeg', mime_content_type($imageFile));
+
+        //$this->assertEquals($imageFile, $media->getPath());
```
Comment (sebastiandedeyne) [inline tests/ImageGenerators/VideoTest.php]: Redundant?

File: src/FileManipulator.php (hunk 2 above)
Comment (freekmurze) [inline src/FileManipulator.php]: much sadness here

File: src/ImageGenerators/BaseGenerator.php (hunk 3 above)
Comment (freekmurze) [inline src/ImageGenerators/BaseGenerator.php]: 👍

File: src/Media.php (hunk 7 above)
Comment (freekmurze) [inline src/Media.php]: 😢

File: tests/ImageGenerators/VideoTest.php (hunk 8 above)
Comment (freekmurze) [inline tests/ImageGenerators/VideoTest.php]: Removed

**Comment by freekmurze** [comment]
@nicolasbeauvais the video test seems to fail: https://travis-ci.org/spatie/laravel-medialibrary/jobs/161723143 

Do you know a good solution to make this pass?

**Comment by nicolasbeauvais** [comment]
I tried to push my fix but couldn't do it :/  I tough GitHub would authorize it as the first commit is mine..

something like this in (https://github.com/spatie/laravel-medialibrary/blob/feature-image-generators/tests/ImageGenerators/VideoTest.php#L23) should do the trick
```php
$imageFile = $imageGenerator->convert($media->getPath(), new Conversion('test'));
```

The video converter needs a conversion object so it can get the `extractVideoAtSecond` result.

**Pull request merged**